### PR TITLE
Simplification and documentation

### PR DESCRIPTION
**Pull request by jensens**
In order to understand whats going on in depth in the zcml directive i took the chance and simplified the code plus I documented it in the README.
Minor accidentially overall cleanup included.

### CODE REVIEW

**Comment by jensens** [comment]
http://jenkins.plone.org/view/All/job/pull-request/115/

**Comment by jensens** [comment]
I missed a MySchemaAwareFactory example... follows soon.

**Comment by jensens** [comment]
Seems like my for check broke a bunch of things, I'll revert this. But the plone.behavior tests are not good if they dont detect this.

File: README.rst [hunk 1]
```diff
@@ -2,39 +2,32 @@
...
 ------------
 
-This package comes with support for registering behaviors and factories. It
-does not, however, implement the policy for determining what behaviors are
-enabled on a particular object at a particular time. That decision is deferred
-to an `IBehaviorAssignable` adapter, which you must implement.
+This package comes with support for registering behaviors and factories.
+It does not, however, implement the policy for determining what behaviors are enabled on a particular object at a particular time.
+That decision is deferred to an ``IBehaviorAssignable`` adapter, which must be implemented.
+I.e. the ``plone.dexterity`` package implements this.
```
Comment (lentinj) [inline README.rst]: Not sure "i.e." is appropriate here. Maybe "plone.dexterity implements this for you"?

**Comment by jensens** [comment]
http://jenkins.plone.org/view/All/job/pull-request/116/

File: README.rst [hunk 2]
```diff
@@ -68,51 +62,162 @@ the type of context being adapted directly. For example::
         def unlock(self):
             # do something
 
-This interface (which describes the type of behavior) and class (which
-describes the implementation of the behavior) then need to be registered.
+This interface (which describes the type of behavior) and class (which describes the implementation of the behavior) then need to be registered.
```
Comment (lentinj) [inline README.rst]: then need*s* to be registered

File: README.rst [hunk 3]
```diff
@@ -68,51 +62,162 @@ the type of context being adapted directly. For example::
...
+``factory``
+    The factory for this behavior (optional).
+    If no factory is given, the behavior context is assumed to provide the interface given by ``provides`` itself.
+
+    If factory provides ``plone.behavior.interfaces.ISchemaAwareFactory`` the factory is assumed to be a callable.
+    It is called with the value given in ``provides`` as the only argument.
+    The value returned is then used as the factory.
+
+    ``plone.behavior.interfaces.ISchemaAwareFactory`` is a marker interface for factories that should be initialised with a schema.
+    Its expected to return another callable that can create appropriate behavior factories on demand.
```
Comment (lentinj) [inline README.rst]: It**s**, or It is. But it seems this paragraph is largely repeating the previous one. Just remove it?

File: README.rst [hunk 4]
```diff
@@ -68,51 +62,162 @@ the type of context being adapted directly. For example::
...
+
+``for``
+    The type of object to register the conditional adapter factory for (optional).
+    Must be skipped is no ``factory`` is given.
```
Comment (lentinj) [inline README.rst]: Must be **omitted** i**f** no factory is given?

**Comment by jensens** [comment]
tests are green and thx @lentinj for your review!

File: plone/behavior/metaconfigure.py [hunk 5]
```diff
@@ -99,10 +88,28 @@ def behaviorDirective(_context, title, provides, description=None, marker=None,
...
+                u"Specifing 'for' in behavior '{0}' if no 'factory' is given "
+                u"has no effect and is superfluos.".format(title)
+            )
+        return
+
+    if for_ is None:
+        # Attempt to guess the factory's adapted interface and use it as
+        # the 'for_'.
+        # Fallback to '*' (=Interface), even if factory is None.
+        adapts = getattr(factory, '__component_adapts__', None) or [Interface]
```
Comment (jensens) [inline plone/behavior/metaconfigure.py]: note to self: factory cant be none here - this is checked some lines above

**Comment by jensens** [comment]
re run tests at http://jenkins.plone.org/job/pull-request/117/

**Comment by jensens** [comment]
Example for ``MySchemaAwareFactory`` and for ``for`` auto-detection added.

I think this one is ready for merge.

**Pull request merged**